Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial.lean

```diff
@@ -0,0 +1,504 @@
+
+@[simp]
+lemma ΨSq_two : W.ΨSq 2 = W.Ψ₂Sq := by
+  erw [ΨSq_ofNat, one_pow, one_mul, if_pos even_two]
+
+@[simp]
+lemma ΨSq_three : W.ΨSq 3 = W.Ψ₃ ^ 2 := by
+  erw [ΨSq_ofNat, Ψ''_three, mul_one]
+
+@[simp]
+lemma ΨSq_four : W.ΨSq 4 = W.Ψ₄' ^ 2 * W.Ψ₂Sq := by
+  erw [ΨSq_ofNat, Ψ''_four, if_pos <| by decide]
+
+lemma ΨSq_odd (m : ℕ) : W.ΨSq (2 * (m + 2) + 1) =
+    (W.Ψ'' (m + 4) * W.Ψ'' (m + 2) ^ 3 * (if Even m then W.Ψ₂Sq ^ 2 else 1) -
+      W.Ψ'' (m + 1) * W.Ψ'' (m + 3) ^ 3 * (if Even m then 1 else W.Ψ₂Sq ^ 2)) ^ 2 := by
+  erw [ΨSq_ofNat, Ψ''_odd, if_neg (m + 2).not_even_two_mul_add_one, mul_one]
+
+lemma ΨSq_even (m : ℕ) : W.ΨSq (2 * (m + 3)) =
+    (W.Ψ'' (m + 2) ^ 2 * W.Ψ'' (m + 3) * W.Ψ'' (m + 5) -
+      W.Ψ'' (m + 1) * W.Ψ'' (m + 3) * W.Ψ'' (m + 4) ^ 2) ^ 2 * W.Ψ₂Sq := by
+  erw [ΨSq_ofNat, Ψ''_even, if_pos <| even_two_mul _]
+
+@[simp]
+lemma ΨSq_neg (n : ℤ) : W.ΨSq (-n) = W.ΨSq n := by
+  rw [WeierstrassCurve.ΨSq, Ψ'_neg, neg_sq, Int.natAbs_neg, WeierstrassCurve.ΨSq]
+
+/-! ### The bivariate polynomials $\Psi_n$ -/
+
+/-- The bivariate polynomials $\Psi_n$ congruent to the $n$-division polynomials $\psi_n$. -/
+protected noncomputable def Ψ (n : ℤ) : R[X][Y] :=
+  C (W.Ψ' n) * if Even n.natAbs then W.ψ₂ else 1
+
+@[simp]
+lemma Ψ_ofNat (n : ℕ) : W.Ψ n = C (W.Ψ'' n) * if Even n then W.ψ₂ else 1 := by
+  rw [WeierstrassCurve.Ψ, Ψ'_ofNat, Int.natAbs_cast]
+
+@[simp]
+lemma Ψ_zero : W.Ψ 0 = 0 := by
+  erw [Ψ_ofNat, C_0, zero_mul]
+
+@[simp]
+lemma Ψ_one : W.Ψ 1 = 1 := by
+  erw [Ψ_ofNat, C_1, mul_one]
+
+@[simp]
+lemma Ψ_two : W.Ψ 2 = W.ψ₂ := by
+  erw [Ψ_ofNat, one_mul, if_pos even_two]
+
+@[simp]
+lemma Ψ_three : W.Ψ 3 = C W.Ψ₃ := by
+  erw [Ψ_ofNat, Ψ''_three, mul_one]
+
+@[simp]
+lemma Ψ_four : W.Ψ 4 = C W.Ψ₄' * W.ψ₂ := by
+  erw [Ψ_ofNat, Ψ''_four, if_pos <| by decide]
+
+lemma Ψ_odd (m : ℕ) : W.Ψ (2 * (m + 2) + 1) =
+    W.Ψ (m + 4) * W.Ψ (m + 2) ^ 3 - W.Ψ (m + 1) * W.Ψ (m + 3) ^ 3 +
+      W.toAffine.polynomial * (16 * W.toAffine.polynomial - 8 * W.ψ₂ ^ 2) * C
+        (if Even m then W.Ψ'' (m + 4) * W.Ψ'' (m + 2) ^ 3
+          else -W.Ψ'' (m + 1) * W.Ψ'' (m + 3) ^ 3) := by
+  repeat erw [Ψ_ofNat]
+  rw [Ψ''_odd, if_neg (m + 2).not_even_two_mul_add_one]
+  by_cases hm : Even m
+  · have hm1 : ¬Even (m + 1) := fun h => Nat.even_add_one.mp h hm
+    have hm2 : Even (m + 2) := Nat.even_add_one.mpr hm1
+    have hm3 : ¬Even (m + 3) := fun h => Nat.even_add_one.mp h hm2
+    have hm4 : Even (m + 4) := Nat.even_add_one.mpr hm3
+    rw [if_pos hm, if_pos hm, if_pos hm4, if_pos hm2, if_neg hm1, if_neg hm3, if_pos hm]
+    C_simp
+    rw [C_Ψ₂Sq_eq]
+    ring1
+  · have hm1 : Even (m + 1) := Nat.even_add_one.mpr hm
+    have hm2 : ¬Even (m + 2) := fun h => Nat.even_add_one.mp h hm1
+    have hm3 : Even (m + 3) := Nat.even_add_one.mpr hm2
+    have hm4 : ¬Even (m + 4) := fun h => Nat.even_add_one.mp h hm3
+    rw [if_neg hm, if_neg hm, if_neg hm4, if_neg hm2, if_pos hm1, if_pos hm3, if_neg hm]
+    C_simp
+    rw [C_Ψ₂Sq_eq]
+    ring1
```

Comment:
You can probably golf the same way as I do [here](https://github.com/leanprover-community/mathlib4/pull/13057/files#diff-02d4668be95cd41b094c2ec5859acf1cf92c74833b9181a533bcce3a23c00961R765-R770).

Reply:
This belongs to #6703 so I will change it there.